app/src/main/java/gholzrib/arctouchchallenge/core/handlers/RequestsHandler.java

```diff
@@ -0,0 +1,153 @@
+package gholzrib.arctouchchallenge.core.handlers;
+
+import android.net.Uri;
+import android.os.AsyncTask;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.util.Map;
+
+import gholzrib.arctouchchallenge.ArcTouchChallengeApplication;
+import gholzrib.arctouchchallenge.core.listeners.RequestsListener;
+import okhttp3.Call;
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+
+public abstract class RequestsHandler {
+
+    public static final MediaType MEDIA_TYPE_JSON = MediaType.parse("application/json; charset=utf-8");
+    public static final MediaType MEDIA_TYPE_JPG = MediaType.parse("image/jpg");
+
+    protected WeakReference<RequestsListener> mApiListener = null;
+    protected WeakReference<Call> mCurrentCall;
+    public int mCurrentOperation;
+
+    public RequestsHandler(RequestsListener requestsListener, int operation) {
+        mApiListener = new WeakReference<>(requestsListener);
+        mCurrentOperation = operation;
+
+        if (mApiListener.get() == null) {
+            throw new IllegalArgumentException("RequestsHandlerListener is required. You mustn't use requestsHandlerListener parameter as null");
+        }
+    }
+
+    protected String createUrl(String baseUrl, String[] paths, Map<String, String> params) {
+        Uri.Builder builder = Uri.parse(baseUrl).buildUpon();
+        if (paths != null) {
+            for (String path : paths) {
+                builder.appendPath(path);
+            }
+        }
+        if (params != null) {
+            for (Map.Entry<String, String> param : params.entrySet()) {
+                builder.appendQueryParameter(param.getKey(), param.getValue());
+            }
+        }
+        return builder.build().toString();
+    }
+
+    public abstract void doRequest(String[] additionalParams, File picture);
+
+    protected void doRequestByUsingGet(final String url) {
+        Request request = new Request.Builder()
+                .url(url)
+                .build();
+        mCurrentCall = new WeakReference<>(ArcTouchChallengeApplication.getClientInstance().newCall(request));
+        ApiTask apiTask = new ApiTask();
+        apiTask.execute(mCurrentCall.get());
+    }
+
+    protected void doRequestByUsingPost(final String url, final String json) {
```

Comment:
Again, this was a class created before the challenge and it's created with the idea to offer support for future features.